pkg/manager/impl/execution_manager.go

```diff
@@ -428,41 +428,77 @@ func (m *ExecutionManager) getInheritedExecMetadata(ctx context.Context, request
 }
 
+// WorkflowExecutionConfigInterface is used as common interface for capturing the common behavior catering to the needs
+// of fetching the WorkflowExecutionConfig across LaunchPlanSpec, ExecutionCreateRequest
+// MatchableResource_WORKFLOW_EXECUTION_CONFIG and ApplicationConfig
+type WorkflowExecutionConfigInterface interface {
+	// GetMaxParallelism Can be used to control the number of parallel nodes to run within the workflow. This is useful to achieve fairness.
+	GetMaxParallelism() int32
+	// GetRawOutputDataConfig Encapsulates user settings pertaining to offloaded data (i.e. Blobs, Schema, query data, etc.).
+	GetRawOutputDataConfig() *admin.RawOutputDataConfig
+	// GetSecurityContext Indicates security context permissions for executions triggered with this matchable attribute.
+	GetSecurityContext() *core.SecurityContext
+	// GetAnnotations Custom annotations to be applied to a triggered execution resource.
+	GetAnnotations() *admin.Annotations
+	// GetLabels Custom labels to be applied to a triggered execution resource.
+	GetLabels() *admin.Labels
+}
+
+// Merge into workflowExecConfig from spec
+func mergeIntoExecConfig(workflowExecConfig *admin.WorkflowExecutionConfig, spec WorkflowExecutionConfigInterface) {
+	if workflowExecConfig.GetMaxParallelism() == 0 && spec.GetMaxParallelism() > 0 {
+		workflowExecConfig.MaxParallelism = spec.GetMaxParallelism()
+	}
+	if workflowExecConfig.GetSecurityContext() == nil && spec.GetSecurityContext() != nil {
+		workflowExecConfig.SecurityContext = spec.GetSecurityContext()
+	}
+	// Launchplan spec has label, annotation and rawOutputDataConfig initialized with empty values.
+	// Hence we do a deep check in the following conditions before assignment
+	if (workflowExecConfig.GetRawOutputDataConfig() == nil ||
+		len(workflowExecConfig.GetRawOutputDataConfig().GetOutputLocationPrefix()) == 0) &&
+		(spec.GetRawOutputDataConfig() != nil && len(spec.GetRawOutputDataConfig().OutputLocationPrefix) > 0) {
+		workflowExecConfig.RawOutputDataConfig = spec.GetRawOutputDataConfig()
+	}
+	if (workflowExecConfig.GetLabels() == nil || len(workflowExecConfig.GetLabels().Values) == 0) &&
+		(spec.GetLabels() != nil && len(spec.GetLabels().Values) > 0) {
+		workflowExecConfig.Labels = spec.GetLabels()
+	}
+	if (workflowExecConfig.GetAnnotations() == nil || len(workflowExecConfig.GetAnnotations().Values) == 0) &&
+		(spec.GetAnnotations() != nil && len(spec.GetAnnotations().Values) > 0) {
+		workflowExecConfig.Annotations = spec.GetAnnotations()
+	}
+}
+
 // Produces execution-time attributes for workflow execution.
 // Defaults to overridable execution values set in the execution create request, then looks at the launch plan values
 // (if any) before defaulting to values set in the matchable resource db and further if matchable resources don't
 // exist then defaults to one set in application configuration
 func (m *ExecutionManager) getExecutionConfig(ctx context.Context, request *admin.ExecutionCreateRequest,
```

Comment:
what if a user specifies some fields in the execution config in the request spec, but not all? 

I think we should then treat that execution config as finalized, and the omitted fields as intentionally excluded.

currently, we keep proceeding with the fallbacks regardless of whether the request spec is populated, or barring that, if the launch plan is then populated. i'm not sure this is right, what do you think? if one of these levels has any non-empty values then should we treat that execution config as the final execution config we apply?

Reply:
I wanted this to look like how we do override for go variables and scopes actually define what that level defines the value for that variable
https://go.dev/play/p/jdkv14fHY8N

But we don't have a way to tell protobuf fields that they are set but those are explicitly being set to nil or empty.
i have seen where some libraries provide a behavior of tristate

i.e set (having value/ having no value) / unset(uninitialized)
Something like this tristateBoolean having isUndef https://docs.oracle.com/cd/E26098_01/apirefs.1112/e17493/oracle/ide/util/TriStateBoolean.html#isUndef__

That way it easier to distinguish what the user wanted to do. 

Also I think in the current state that issue comes even for maxParallelism = 0 in execution create request  and launch plan spec having the same value and  it will fall back to other overrides though the user intended to set maxParallelism = 0 in execution create Request 

Having a tristate behavior solves such cases and make it easier to reason across overrides and avoid using custom logic IMO.

To keep logic similar to previous one , i am adding a isChanged bool flag in merge function which is checked after the merge from execution create request and launchplan spec and if it gets set then we return that config object.

Let me know what you think.

Reply:
re: the tri-state, we do something funky with one-ofs for notification overrides to permit a user to explicitly [disable notifications](https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/admin/execution.proto#L263) as opposed to interpreting the optional field as omitted. we could consider updating the execution config in the proto definitions to have some kind of bool flag for "is set" so that even an empty execution config is considered explicitly set

isChanged looks good for now too!